cohortextractor/dsl.py

```diff
@@ -0,0 +1,309 @@
+for end users.
+"""
+
+from __future__ import annotations
+
+from typing import Any
+
+from cohortextractor.query_language import BaseTable as QMTable
+from cohortextractor.query_language import (
+    QueryNode,
+    Row,
+    Table,
+    Value,
+    ValueFromAggregate,
+    ValueFromRow,
+)
+
+
+class Cohort:
+    """Represents the cohort of patients in a study."""
+
+    def set_population(self, population: PatientSeries) -> None:
+        """Set the population variable for this cohort."""
+
+    def add_variable(self, name: str, variable: PatientSeries) -> None:
+        """Add a variable to this cohort by name."""
+        object.__setattr__(self, name, variable)
+
+    def __setattr__(self, name: str, variable: PatientSeries) -> None:
+        return self.add_variable(name, variable)
+
+
+class Node:
+    name: str
+
+    def __init__(
+        self,
+        parent: Node,
+        method_name: str,
+        *args: Any,
+        **kwargs: Any,
+    ):
+        self._path: list[PathItem] = parent._path + [
+            (parent, method_name, args, kwargs)
+        ]
+
+
+PathItem = tuple[Node, str, Any, Any]
+State = dict[str, Any]
+
+
+class EventFrame(Node):
+    """Represents a collection of records, with multiple rows per patient.
+
+    Either an EventTable, or the result of filtering an EventTable.
+    """
+
+    def filter(self, column: str, **kwargs: str) -> EventFrame:  # noqa: A003
+        """Return a new EventFrame with given filter."""
+
+        return EventFrame(self, "filter", column, **kwargs)
+
+    def sort_by(self, *columns: str) -> SortedEventFrame:
+        """Return a SortedEventFrame with given sort column."""
+
+        return SortedEventFrame(self, "sort_by", *columns)
+
+    def count_for_patient(self) -> PatientSeries:
+        """Return a PatientSeries with count_for_patient of matching events per patient."""
+
+        return PatientSeries(self, "count_for_patient")
+
+    def exists_for_patient(self) -> PatientSeries:
+        """Return a PatientSeries indicating whether each patient has a matching event."""
+
+        return PatientSeries(self, "exists_for_patient")
+
+    def _filter(
+        self,
+        table: QMTable,
+        state: State,
+        column: str,
+        **kwargs: str,
+    ) -> tuple[QMTable, State]:
+        if kwargs.get("not_equals", object()) is None:
+            assert "not_null_filter" not in state
+            extra_state = {"not_null_filter": column}
+        else:
+            extra_state = {}
+        return table.filter(column, **kwargs), extra_state
+
+    def _sort_by(
+        self,
+        table: QMTable,
+        state: State,
+        *columns: str,
+    ) -> tuple[QMTable, State]:
+        return table, {"sort_columns": columns}
+
+    def _exists_for_patient(
+        self,
+        table: QMTable,
+        state: State,
+    ) -> tuple[ValueFromAggregate, State]:
+        return table.exists(state["not_null_filter"]), {}
```

Comment:
Can't we call `exists` without filtering first (so there'd be no `not_null_filter` in state?

Reply:
Yes, good point.